File: src/use-cases/auth/RegisterUser.js
/**
 * RegisterUser Use Case
 * Handles user registration with validation and JWT token generation
 */

const User = require('../../domain/entities/User');
const logger = require('../../../logger');

class RegisterUser {
    /**
     * @param {IUserRepository} userRepository - User repository
     * @param {JwtService} jwtService - JWT service
     */
    constructor(userRepository, jwtService) {
        this.userRepository = userRepository;
        this.jwtService = jwtService;
    }

    /**
     * Validate username format
     * @param {string} username - Username to validate
     * @returns {{valid: boolean, message: string|null}}
     */
    validateUsername(username) {
        if (!username || typeof username !== 'string') {
            return { valid: false, message: 'Le pseudo est requis' };
        }

        const trimmed = username.trim();

        if (trimmed.length < 3) {
            return { valid: false, message: 'Le pseudo doit contenir au moins 3 caractères' };
        }

        if (trimmed.length > 30) {
            return { valid: false, message: 'Le pseudo ne peut pas dépasser 30 caractères' };
        }

        // Only allow letters, numbers, hyphens and underscores
        const validUsernameRegex = /^[a-zA-Z0-9_-]+$/;
        if (!validUsernameRegex.test(trimmed)) {
            return {
                valid: false,
                message: 'Le pseudo ne peut contenir que des lettres, chiffres, tirets (-) et underscores (_)'
            };
        }

        return { valid: true, message: null };
    }

    /**
     * Validate password format
     * @param {string} password - Password to validate
     * @returns {{valid: boolean, message: string|null}}
     */
    validatePassword(password) {
        if (!password || typeof password !== 'string') {
            return { valid: false, message: 'Le mot de passe est requis' };
        }

        if (password.length < 6) {
            return { valid: false, message: 'Le mot de passe doit contenir au moins 6 caractères' };
        }

        if (password.length > 100) {
            return { valid: false, message: 'Le mot de passe ne peut pas dépasser 100 caractères' };
        }

        return { valid: true, message: null };
    }

    /**
     * Execute the use case
     * @param {Object} request - Registration request
     * @param {string} request.username - Username (3-30 chars, alphanumeric with - and _)
     * @param {string} request.password - Password (6-100 chars)
     * @returns {Promise<Object>} Registration result with user and token
     */
    async execute({ username, password }) {
        try {
            // Validate username
            const usernameValidation = this.validateUsername(username);
            if (!usernameValidation.valid) {
                throw new Error(usernameValidation.message);
            }

            // Validate password
            const passwordValidation = this.validatePassword(password);
            if (!passwordValidation.valid) {
                throw new Error(passwordValidation.message);
            }

            const trimmedUsername = username.trim();

            // Check if username already exists
            const existingUser = await this.userRepository.existsByUsername(trimmedUsername);

            if (existingUser) {
                logger.warn('Registration failed: username already exists', { username: trimmedUsername });
                throw new Error('Ce pseudo est déjà pris');
            }

            // Create new user (will hash password)
            const user = await User.create(trimmedUsername, password);

            // Save to repository
            await this.userRepository.save(user);

            // Generate JWT token
            const token = this.jwtService.sign({
                userId: user.id,
                username: user.username
            });

            logger.info('User registered successfully', {
                userId: user.id,
                username: user.username
            });

            return {
                success: true,
                user: user.toPublicObject(),
                token
            };
        } catch (error) {
            logger.error('Registration error', {
                username,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = RegisterUser;

File: zapzap-rust/src/application/auth/register_user.rs
use std::sync::Arc;

use uuid::Uuid;

use crate::domain::entities::User;
use crate::domain::repositories::UserRepository;
use crate::infrastructure::auth::{JwtService, PasswordService};

/// Register user input
pub struct RegisterUserInput {
    pub username: String,
    pub password: String,
}

/// Register user output
pub struct RegisterUserOutput {
    pub user: User,
    pub token: String,
}

/// Register user use case
pub struct RegisterUser {
    user_repo: Arc<dyn UserRepository>,
    jwt_service: Arc<JwtService>,
}

impl RegisterUser {
    pub fn new(user_repo: Arc<dyn UserRepository>, jwt_service: Arc<JwtService>) -> Self {
        Self {
            user_repo,
            jwt_service,
        }
    }

    pub async fn execute(&self, input: RegisterUserInput) -> Result<RegisterUserOutput, RegisterError> {
        // Validate input
        if input.username.trim().is_empty() {
            return Err(RegisterError::Validation("Username is required".into()));
        }
        if input.username.len() < 3 {
            return Err(RegisterError::Validation("Username must be at least 3 characters".into()));
        }
        if input.password.len() < 4 {
            return Err(RegisterError::Validation("Password must be at least 4 characters".into()));
        }

        // Check if username exists
        if self.user_repo.exists_by_username(&input.username).await? {
            return Err(RegisterError::UsernameExists);
        }

        // Hash password
        let password_hash = PasswordService::hash(&input.password)
            .map_err(|e| RegisterError::Internal(e.to_string()))?;

        // Create user
        let user_id = Uuid::new_v4().to_string();
        let user = User::new_human(user_id, input.username.clone(), password_hash);

        // Save user
        self.user_repo.save(&user).await?;

        // Generate token
        let token = self
            .jwt_service
            .sign(&user.id, &user.username, user.is_admin)
            .map_err(|e| RegisterError::Internal(e.to_string()))?;

        Ok(RegisterUserOutput { user, token })
    }
}

/// Register error types
#[derive(Debug, thiserror::Error)]
pub enum RegisterError {
    #[error("Validation error: {0}")]
    Validation(String),
    #[error("Username already exists")]
    UsernameExists,
    #[error("Internal error: {0}")]
    Internal(String),
    #[error("Repository error: {0}")]
    Repository(#[from] crate::domain::repositories::RepositoryError),
}
